components/fxa-client/android/src/main/java/mozilla/appservices/fxa/FirefoxAccount.kt

```diff
@@ -0,0 +1,153 @@
+     */
+    abstract suspend fun getAccessToken(singleScope: String): AccessTokenInfo?
+
+    /**
+     * Fetches the profile object for the current client either from the existing cached state
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return Profile (optional, if successfully retrieved) representing the user's basic profile info
+     */
+    abstract suspend fun getProfile(ignoreCache: Boolean = false): Profile?
+
+    /**
+     * Fetches the token server endpoint, for authentication using the SAML bearer flow.
+     *
+     * @return Token server endpoint URL string, `null` if it couldn't be obtained.
+     */
+    abstract suspend fun getTokenServerEndpointURL(): String?
+
+    /**
+     * Get the pairing URL to navigate to on the Authority side (typically a computer).
+     *
+     * @return The URL to show the pairing user
+     */
+    abstract fun getPairingAuthorityURL(): String
+
+    /**
+     * Returns current FxA Device ID for an authenticated account.
+     *
+     * @return Current device's FxA ID, if available. `null` otherwise.
+     */
+    abstract fun getCurrentDeviceId(): String?
+
+    /**
+     * Returns session token for an authenticated account.
+     *
+     * @return Current account's session token, if available. `null` otherwise.
+     */
+    abstract fun getSessionToken(): String?
+
+    /**
+     * Indicates if account needs to be re-authenticated via [beginAuthentication].
+     * Most common reason for an account to need re-authentication is a password change.
+     *
+     * TODO this may return a false-positive, if we're currently going through a recovery flow.
+     * Prefer to be notified of auth problems via [AccountObserver], which is reliable.
+     *
+     * @return A boolean flag indicating if account needs to be re-authenticated.
+     */
+    abstract fun accountNeedsReauth(): Boolean
+
+    /**
+     * This method should be called when a request made with an OAuth token failed with an
+     * authentication error. It will re-build cached state and perform a connectivity check.
+     *
+     * In time, fxalib will grow a similar method, at which point we'll just relay to it.
+     * See https://github.com/mozilla/application-services/issues/1263
+     *
+     * @param singleScope An oauth scope for which to check authorization state.
+     * @return An optional [Boolean] flag indicating if we're connected, or need to go through
+     * re-authentication. A null result means we were not able to determine state at this time.
+     */
+    abstract suspend fun checkAuthorizationStatus(singleScope: String): Boolean?
+
+    /**
+     * Register for account events
+     */
+    abstract suspend fun registerAccountEventHandler(handler: AccountEventHandler)
+
+    abstract fun stateManager(): StateManager
+
+    abstract fun deviceManager(): DeviceManager
```

Comment:
One thing I liked about the android components code was that it split things up into different components, which helped keep things understandable.  What do you think of doing this for our component?  What is the correct split?

Reply:
Which part of a-c are you referring to? I still don't really understand some of the splits in a-c - I can't tell you how a `GlobalAccountManager` relates to an `FxaAccountManager`, and `SyncEnginesStorage`, `AccountStorage` and `SyncAuthInfoCache` all seems vaguely related to storage but aren't in a directory names `store`?

Reply:
It's confusing to me as well and I think there's way too many files overall.  But I do like the general idea of splitting up the API surface.  The current logic of this PR goes like this:
  - `FirefoxAccount` is kind of like [FxaAccountManager](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt).  It's the top-level class and handles general functionality
  - `DeviceManager` is kind of like [FxaDeviceConstellation](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceConstellation.kt).  It handles device-related functionality.
  - `StateManager` is kind of like [FirefoxAccount](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FirefoxAccount.kt) it has methods to initiate state transitions.  This one has a lot of changes compared to the android-components class. I made it focus only on state transitions, which meant removing methods like `getAccessToken()`, and I also added the `Uninitialized` state.